protocols/gossipsub/src/handler.rs

```diff
@@ -49,12 +76,34 @@ pub struct GossipsubHandler {
     inbound_substream: Option<InboundSubstreamState>,
 
     /// Queue of values that we want to send to the remote.
-    send_queue: SmallVec<[GossipsubRpc; 16]>,
+    send_queue: SmallVec<[crate::rpc_proto::Rpc; 16]>,
 
     /// Flag indicating that an outbound substream is being established to prevent duplicate
     /// requests.
     outbound_substream_establishing: bool,
 
+    /// The number of outbound substreams we have created.
+    outbound_substreams_created: usize,
+
+    /// The number of inbound substreams that have been created by the peer.
+    inbound_substreams_created: usize,
+
+    /// The type of peer this handler is associated to.
+    peer_kind: Option<PeerKind>,
+
+    /// Keeps track on whether we have sent the peer kind to the behaviour.
+    // NOTE: Use this flag rather than checking the substream count each poll.
+    peer_kind_sent: bool,
+
+    /// If the peer doesn't support the gossipsub protocol we do not immediately disconnect.
```

Comment:
Can you explain why the handler is kept alive even though gossipsub does not utilize the underlying connection?

Reply:
IIRC, the previous functionality is such that if a peer does not support gossipsub or floodsub it would be immediately disconnected. 

This allows the handler to not disconnect but rather set the KeepAlive to false. 

I think this is important for more complex NetworkBehaviours/ProtocolHandlers that may allow for other protocols and may wish to keep connections with peers that don't support gossipsub but may support another behaviour. In this way, the global network behaviour can drop peers based on keep alive and have the option to drop peers based on their supported protocols.